### ISSUE TITLE
cuda inference time is similar to cpu

### ISSUE THREAD

**Issue by yanisIk** [ultralytics/yolov5#5461]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection, PyTorch Hub

### Bug

Getting the exact same performance when using `model.device = 'cuda' or 'cpu'`

I'm getting ~20 fps using yoloV5m and ~30 fps using yoloV5s, by setting `cuda` or `cpu`

All my drivers and dependencies are properly installed following the official PyTorch and Nvidia documentation and verified with `torch.cuda.is_available()` returning `True`

### Environment

- YOLOv5  2021-11-1 torch 1.10.0+cu113 CUDA:0 (NVIDIA GeForce RTX 3080, 10239.375MB)
- OS: Windows 11
- Python 3.8
- Cuda and CuDNN properly installed
- `torch.cuda.is_available()` returns `True`

### Minimal Reproducible Example

```
def load_model(device):
    model = torch.hub.load('ultralytics/yolov5', 'yolov5m')
    model.device = device
    return model

def detect(model, img):
    results = model(img)
    results_df = results.pandas().xyxy[0]
    results_df = results_df.sort_values(by=['confidence'], ascending=False).head(1)
    return results_df


model = load_model(device='cuda') # Changing it to cpu gives the same performance
images = [np.array(...), ...] # list of raw images in numpy arrays (640*640)

for img in images:
    start_time = timeit.default_timer()
    detect(model, img)
    end_time = timeit.default_timer()
    fps = int(1 / (end_time - start_time))
    print(fps)
```

### Additional

yolo5m takes 4gb of ram when loaded, is this a proof that it's using CPU and not GPU ? (shouldn't it be in the VRAM ?)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@yanisIk you may have environment problems. You can run an example of correct GPU/CPU inference speeds in our Colab Notebook:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# GPU Inference
model.cuda()
results = model(imgs)  # skip one
results = model(imgs)  # profile 
results.print()  # or .show(), .save()
# Speed: 3.1ms pre-process, 6.7ms inference, 1.5ms NMS per image at shape (2, 3, 640, 640)

# CPU Inference
model.cpu()
results = model(imgs)  # skip one
results = model(imgs)  # profile 
results.print()  # or .show(), .save()
# Speed: 6.0ms pre-process, 369.7ms inference, 1.0ms NMS per image at shape (2, 3, 640, 640)
```

**Comment by iceisfun** [comment]
@yanisIk the critical missing step here is probably the warmup

i do something like
```
model([ np.zeros(shape=[720, 1280, 3], dtype=np.uint8) ])
```

the reference code uses the first image
```
results = model(imgs)  # skip one
```

**Comment by glenn-jocher** [comment]
@yanisIk @iceisfun yes CUDA inference is always slow on the first image, so for proper profiling you should run inference once to warm up, and then all other batches afterward are fast.

**Comment by yanisIk** [comment]
Thanks a lot for your answers ! I didn't know about that and didn't know about the `.cuda()` and `.cpu()` methods.

@glenn-jocher I ran your test script on my machine and on google collab and got the following results:

- **CUDA**: Speed: 3.5ms pre-process, 31.5ms inference, 2.0ms NMS per image at shape (2, 3, 640, 640)
- **CPU**: Speed: 3.5ms pre-process, 222.5ms inference, 1.5ms NMS per image at shape (2, 3, 640, 640)
- **Colab CUDA**: Speed: 4.1ms pre-process, 36.4ms inference, 1.8ms NMS per image at shape (2, 3, 640, 640)

Now i can definitely see the difference between CUDA and CPU (which didn't happen when using `model.device = ...` )

### **Now my question would be:**

Why am I getting 31.5ms inference speed using the small model on an RTX 3080 ?
How can I achieve sub 10ms inference time like in your benchmarks ?

**Comment by glenn-jocher** [comment]
@yanisIk I don't understand your expectations for matching README results, which are based on V100 GCP instances locally on completely different GPU and hardware. The answer to your question of how to reproduce the README results is to run the README reproduction commands on the same hardware, otherwise you're asking why an apple is not an orange.


## About Reported Speeds

mAP values are reproducible across any hardware, but **speeds will vary significantly among V100 instances**, and seem to depend heavily on the CUDA, CUDNN and PyTorch installations used. The numbers reported above were produced on GCP [N1-standard-8](https://cloud.google.com/compute/docs/machine-types) Skylake V100 instances running the [v6.0 Docker image](https://hub.docker.com/layers/ultralytics/yolov5/v6.0/images/sha256-58fe76178d82740e5e1c87b382590c7eca5beb952c8b7bab50b5cadeac6eb967?context=explore) with:

- NVIDIA Driver Version: 460.73.01
- CUDA Version: 11.2 

```bash
# Pull and Run v6.0 image
t=ultralytics/yolov5:v6.0 && sudo docker pull $t && sudo docker run -it --ipc=host --gpus all $t
```

<img width="1048" alt="Screen Shot 2021-10-12 at 10 57 31 AM" src="https://user-images.githubusercontent.com/26833433/137006376-a0d8eba0-fe50-4110-b953-69f7b3eeb08d.png">

Our speed command is:
```bash
# Reproduce YOLOv5s batch-1 speeds in table
python val.py --data coco.yaml --img 640 --task speed --batch 1
```
<img width="1214" alt="Screen Shot 2021-10-12 at 11 03 34 AM" src="https://user-images.githubusercontent.com/26833433/137006804-b21c7391-af9b-4867-a41d-60a89235af9b.png">

We tried several options, including AWS P3 instances, pulling more recent base versions (`FROM nvcr.io/nvidia/pytorch:21.09-py3`), different pytorch install methods, and many returned slower speeds than the values reported in our table. batch-32 speeds were found to vary less across options.

**Comment by yanisIk** [comment]
@glenn-jocher Thanks for the quick response.

I just wanted to make sure that my system was working properly, especially that I'm getting the same speed from a K80 (collab) and an RTX 3080 and have no other reference point.

What would be the best way to check if someone's system is behaving correctly ? Is there a table somewhere that shows what we should expect based on hardware ? (That would be maybe better to ask this on a different issue ?)

Also one quick question: Which yolo5 model would be "equivalent" to yolov3-tiny ?

Thanks a lot

**Comment by glenn-jocher** [comment]
@yanisIk there's not really any equivalent YOLOv3-tiny model in YOLOv5, though YOLOv5n may be the closest. 

Your results will vary by hardware, software, drivers, and versions of packages installed. As the above comments explain the README results are reproducible using that exact setup, you can't draw conclusions about any other setup, you must simply evaluate and experiment to see what works best for your situation.

**Comment by yanisIk** [comment]
@glenn-jocher 

Thank you for your help, this issue can be closed now.

I will do more extra research on my side to compare performances with similar systems.
I'm looking to achieve 60fps detection, I've seen people achieving 1000+ fps using RTX 2080 with yolov3-tiny and openCV cudnn, I'm just now trying to get ~60fps with my 3080, which i can't get even using the nano model (~35 - 40 fps).

**Comment by glenn-jocher** [comment]
@yanisIk I think you're confusing results. The fastest results are at large batch sizes. As I've said multiple times the commands to reproduce are in the README, and clearly show batch size. I would suggest running them.
